lib/spack/spack/build_environment.py

```diff
@@ -255,25 +347,38 @@ def set_compiler_environment_variables(pkg, env):
     assert all(key in compiler.link_paths for key in (
         'cc', 'cxx', 'f77', 'fc'))
 
+    wrapper_driver = ''
+    # Git Bash requires this to avoid mangling all paths/forward slashes.
+    # Git bash normally performs this posix mangling: https://github.com/msys2/path_convert
+    # If there are inexplicable errors with compilers interpreting
+    # command line arguments as paths on Windows, check here first
+    # as this variable is subject to change as Git Bash updates
+    if is_windows:
+        env.set('MSYS_NO_PATHCONV', 1)
+        wrapper_driver = "'C:\\Program Files\\Git\\bin\\sh.exe'"
+
     # Populate an object with the list of environment modifications
     # and return it
     # TODO : add additional kwargs for better diagnostics, like requestor,
     # ttyout, ttyerr, etc.
     link_dir = spack.paths.build_env_path
 
+    join = posixpath.join
+    gen_compiler_str = lambda x: "{} {}".format(wrapper_driver, join(link_dir, x))
```

Comment:
(question) You mention doing this in the description:

> Spack now prepends all compiler wrapper environment/CL settings with a call to a users Git sh.

I'm curious why that's needed (this is more about my ignorance of how things work on Windows). 

Is it possible to build packages with https://github.com/spack/spack/pull/27021 without these changes?

Reply:
It is possible to build packages in the primary windows support feature branch without these changes due to an override of the `CC`, `CXX`, `FC`, etc env variables in  Spack `compiler/msvc.py::MSVC::setup_environment` to the actual compiler paths on Windows. Basically we just ignored the concept of the wrappers to begin with.

Invoking Git's vendored copy of sh allows Windows to execute the compiler wrapper shell scripts because Windows does not derive the concept of an executable from the file header itself, but instead queries the registry to determine if and how to execute a given file based on its extension (or lack thereof). It's very possible to, rather than prepend calls to the compiler wrappers with Git sh, to just edit the registry to teach Windows to execute the compiler wrappers (which are extensionless, another small issue for Windows) with Git sh. That's the approach I initially took. However this solution was far from portable, and because Windows sometimes (based on system settings) requires administrative privileges to access some of the changes that need to be made, it didn't work on some systems. Additionally if a user has already set the registry up to execute a certain file extension a certain way, there is no way for Spack to override that without admin privs. Essentially, prepending the path to Git sh is a far more robust solution, if a little bizarre at first.